depends/patches/libsnark/2_unsafe_sha256_speedup.patch

```diff
@@ -0,0 +1,377 @@
+-    round_c.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 5*32, prev_output.rbegin() + 6*32));
+-    round_d.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 4*32, prev_output.rbegin() + 5*32));
+-    round_e.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 3*32, prev_output.rbegin() + 4*32));
+-    round_f.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 2*32, prev_output.rbegin() + 3*32));
+-    round_g.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 1*32, prev_output.rbegin() + 2*32));
+-    round_h.push_back(pb_linear_combination_array<FieldT>(prev_output.rbegin() + 0*32, prev_output.rbegin() + 1*32));
+-
+-    /* do the rounds */
+-    for (size_t i = 0; i < 64; ++i)
+-    {
+-        round_h.push_back(round_g[i]);
+-        round_g.push_back(round_f[i]);
+-        round_f.push_back(round_e[i]);
+-        round_d.push_back(round_c[i]);
+-        round_c.push_back(round_b[i]);
+-        round_b.push_back(round_a[i]);
+-
+-        pb_variable_array<FieldT> new_round_a_variables;
+-        new_round_a_variables.allocate(pb, 32, FMT(this->annotation_prefix, " new_round_a_variables_%zu", i+1));
+-        round_a.emplace_back(new_round_a_variables);
+-
+-        pb_variable_array<FieldT> new_round_e_variables;
+-        new_round_e_variables.allocate(pb, 32, FMT(this->annotation_prefix, " new_round_e_variables_%zu", i+1));
+-        round_e.emplace_back(new_round_e_variables);
+-
+-        round_functions.push_back(sha256_round_function_gadget<FieldT>(pb,
+-                                                                       round_a[i], round_b[i], round_c[i], round_d[i],
+-                                                                       round_e[i], round_f[i], round_g[i], round_h[i],
+-                                                                       packed_W[i], SHA256_K[i], round_a[i+1], round_e[i+1],
+-                                                                       FMT(this->annotation_prefix, " round_functions_%zu", i)));
+-    }
+-
+-    /* finalize */
+-    unreduced_output.allocate(pb, 8, FMT(this->annotation_prefix, " unreduced_output"));
+-    reduced_output.allocate(pb, 8, FMT(this->annotation_prefix, " reduced_output"));
+-    for (size_t i = 0; i < 8; ++i)
+-    {
+-        reduce_output.push_back(lastbits_gadget<FieldT>(pb,
+-                                                        unreduced_output[i],
+-                                                        32+1,
+-                                                        reduced_output[i],
+-                                                        pb_variable_array<FieldT>(output.bits.rbegin() + (7-i) * 32, output.bits.rbegin() + (8-i) * 32),
+-                                                        FMT(this->annotation_prefix, " reduce_output_%zu", i)));
++    if (!pb.unsafe_speedup) {
```

Comment:
This would be better expressed as
```
if (pb.unsafe_speedup) {
    ...
    return;
}
```
to avoid the large diff in the libsnark code.